File: app/src/main/java/dev/ragnarok/fenrir/mvp/presenter/wallattachments/WallPostQueryAttachmentsPresenter.java
package dev.ragnarok.fenrir.mvp.presenter.wallattachments;

import static dev.ragnarok.fenrir.util.RxUtils.dummy;
import static dev.ragnarok.fenrir.util.RxUtils.ignore;
import static dev.ragnarok.fenrir.util.Utils.getCauseIfRuntime;
import static dev.ragnarok.fenrir.util.Utils.safeCountOf;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.api.model.VKApiPost;
import dev.ragnarok.fenrir.domain.IWallsRepository;
import dev.ragnarok.fenrir.domain.Repository;
import dev.ragnarok.fenrir.model.Article;
import dev.ragnarok.fenrir.model.Document;
import dev.ragnarok.fenrir.model.Link;
import dev.ragnarok.fenrir.model.Photo;
import dev.ragnarok.fenrir.model.PhotoAlbum;
import dev.ragnarok.fenrir.model.Poll;
import dev.ragnarok.fenrir.model.Post;
import dev.ragnarok.fenrir.model.Video;
import dev.ragnarok.fenrir.model.criteria.WallCriteria;
import dev.ragnarok.fenrir.mvp.presenter.base.PlaceSupportPresenter;
import dev.ragnarok.fenrir.mvp.view.wallattachments.IWallPostQueryAttachmentsView;
import dev.ragnarok.fenrir.util.RxUtils;
import dev.ragnarok.fenrir.util.Utils;
import io.reactivex.rxjava3.disposables.CompositeDisposable;

public class WallPostQueryAttachmentsPresenter extends PlaceSupportPresenter<IWallPostQueryAttachmentsView> {

    private final ArrayList<Post> mPost;
    private final IWallsRepository fInteractor;
    private final int owner_id;
    private final CompositeDisposable actualDataDisposable = new CompositeDisposable();
    private int loaded;
    private boolean actualDataReceived;
    private boolean endOfContent;
    private boolean actualDataLoading;
    private String Query;

    public WallPostQueryAttachmentsPresenter(int accountId, int ownerId, @Nullable Bundle savedInstanceState) {
        super(accountId, savedInstanceState);
        owner_id = ownerId;
        mPost = new ArrayList<>();
        fInteractor = Repository.INSTANCE.getWalls();
    }

    @Override
    public void onGuiCreated(@NonNull IWallPostQueryAttachmentsView view) {
        super.onGuiCreated(view);
        view.displayData(mPost);

        resolveToolbar();
    }

    private void loadActualData(int offset) {
        actualDataLoading = true;

        resolveRefreshingView();

        int accountId = getAccountId();
        actualDataDisposable.add(fInteractor.getWallNoCache(accountId, owner_id, offset, 100, WallCriteria.MODE_ALL)
                .compose(RxUtils.applySingleIOToMainSchedulers())
                .subscribe(data -> onActualDataReceived(offset, data), this::onActualDataGetError));

    }

    public void fireSearchRequestChanged(String q, boolean only_insert) {
        Query = q == null ? null : q.trim();
        if (only_insert) {
            return;
        }
        actualDataDisposable.clear();
        actualDataLoading = false;
        resolveRefreshingView();
        callView(v -> v.onSetLoadingStatus(0));
        fireRefresh();
    }

    private void onActualDataGetError(Throwable t) {
        actualDataLoading = false;
        callView(v -> showError(v, getCauseIfRuntime(t)));

        resolveRefreshingView();
    }

    private boolean check(String data, List<String> str) {
        for (String i : str) {
            if (data.toLowerCase().contains(i)) {
                return true;
            }
        }
        return false;
    }

    private boolean doCompare(String data, List<String> str) {
        return Utils.safeCheck(data, () -> check(data, str));
    }

    private boolean checkDocs(ArrayList<Document> docs, List<String> str, List<Integer> ids) {
        if (Utils.isEmpty(docs)) {
            return false;
        }
        for (Document i : docs) {
            if (doCompare(i.getTitle(), str) || doCompare(i.getExt(), str) || ids.contains(i.getOwnerId())) {
                return true;
            }
        }
        return false;
    }

    private boolean checkPhotos(ArrayList<Photo> docs, List<String> str, List<Integer> ids) {
        if (Utils.isEmpty(docs)) {
            return false;
        }
        for (Photo i : docs) {
            if (doCompare(i.getText(), str) || ids.contains(i.getOwnerId())) {
                return true;
            }
        }
        return false;
    }

    private boolean checkVideos(ArrayList<Video> docs, List<String> str, List<Integer> ids) {
        if (Utils.isEmpty(docs)) {
            return false;
        }
        for (Video i : docs) {
            if (doCompare(i.getTitle(), str) || doCompare(i.getDescription(), str) || ids.contains(i.getOwnerId())) {
                return true;
            }
        }
        return false;
    }

    private boolean checkAlbums(ArrayList<PhotoAlbum> docs, List<String> str, List<Integer> ids) {
        if (Utils.isEmpty(docs)) {
            return false;
        }
        for (PhotoAlbum i : docs) {
            if (doCompare(i.getTitle(), str) || doCompare(i.getDescription(), str) || ids.contains(i.getOwnerId())) {
                return true;
            }
        }
        return false;
    }

    private boolean checkLinks(ArrayList<Link> docs, List<String> str) {
        if (Utils.isEmpty(docs)) {
            return false;
        }
        for (Link i : docs) {
            if (doCompare(i.getTitle(), str) || doCompare(i.getDescription(), str) || doCompare(i.getCaption(), str)) {
                return true;
            }
        }
        return false;
    }

    private boolean checkArticles(ArrayList<Article> docs, List<String> str, List<Integer> ids) {
        if (Utils.isEmpty(docs)) {
            return false;
        }
        for (Article i : docs) {
            if (doCompare(i.getTitle(), str) || doCompare(i.getSubTitle(), str) || ids.contains(i.getOwnerId())) {
                return true;
            }
        }
        return false;
    }

    private boolean checkPoll(ArrayList<Poll> docs, List<String> str, List<Integer> ids) {
        if (Utils.isEmpty(docs)) {
            return false;
        }
        for (Poll i : docs) {
            if (doCompare(i.getQuestion(), str) || ids.contains(i.getOwnerId())) {
                return true;
            }
        }
        return false;
    }

    private void update(List<Post> data, List<String> str, List<Integer> ids) {

        for (Post i : data) {
            if ((i.hasText() && doCompare(i.getText(), str)) || ids.contains(i.getOwnerId()) || ids.contains(i.getSignerId()) || ids.contains(i.getAuthorId())) {
                mPost.add(i);
            } else if (i.getAuthor() != null && doCompare(i.getAuthor().getFullName(), str)) {
                mPost.add(i);
            } else if (i.hasAttachments() && (checkDocs(i.getAttachments().getDocs(), str, ids)
                    || checkAlbums(i.getAttachments().getPhotoAlbums(), str, ids) || checkArticles(i.getAttachments().getArticles(), str, ids)
                    || checkLinks(i.getAttachments().getLinks(), str) || checkPhotos(i.getAttachments().getPhotos(), str, ids)
                    || checkVideos(i.getAttachments().getVideos(), str, ids) || checkPoll(i.getAttachments().getPolls(), str, ids))) {
                mPost.add(i);
            }
            if (i.hasCopyHierarchy())
                update(i.getCopyHierarchy(), str, ids);
        }
    }

    private void onActualDataReceived(int offset, List<Post> data) {

        actualDataLoading = false;
        endOfContent = data.isEmpty();
        actualDataReceived = true;
        if (endOfContent)
            callResumedView(v -> v.onSetLoadingStatus(2));

        String[] str = Query.split("\\|");
        for (int i = 0; i < str.length; i++) {
            str[i] = str[i].trim().toLowerCase();
        }

        List<Integer> ids = new ArrayList<>();
        for (String cc : str) {
            if (cc.contains("*id")) {
                try {
                    ids.add(Integer.parseInt(cc.replace("*id", "")));
                } catch (NumberFormatException ignored) {
                }
            }
        }

        if (offset == 0) {
            loaded = data.size();
            mPost.clear();
            update(data, Arrays.asList(str), ids);
            resolveToolbar();
            callView(IWallPostQueryAttachmentsView::notifyDataSetChanged);
        } else {
            int startSize = mPost.size();
            loaded += data.size();
            update(data, Arrays.asList(str), ids);
            resolveToolbar();
            callView(view -> view.notifyDataAdded(startSize, mPost.size() - startSize));
        }

        resolveRefreshingView();
    }

    @Override
    public void onGuiResumed() {
        super.onGuiResumed();
        resolveRefreshingView();
    }

    private void resolveRefreshingView() {
        callResumedView(v -> v.showRefreshing(actualDataLoading));
        if (!endOfContent)
            callResumedView(v -> v.onSetLoadingStatus(actualDataLoading ? 1 : 0));
    }

    private void resolveToolbar() {
        callView(v -> {
            v.setToolbarTitle(getString(R.string.attachments_in_wall));
            v.setToolbarSubtitle(getString(R.string.query, safeCountOf(mPost)) + " " + getString(R.string.posts_analized, loaded));
        });
    }

    @Override
    public void onDestroyed() {
        actualDataDisposable.dispose();
        super.onDestroyed();
    }

    public boolean fireScrollToEnd() {
        if (Utils.isEmpty(Query)) {
            return true;
        }
        if (!endOfContent && actualDataReceived && !actualDataLoading) {
            loadActualData(loaded);
            return false;
        }
        return true;
    }

    public void fireRefresh() {
        if (Utils.isEmpty(Query)) {
            return;
        }
        actualDataDisposable.clear();
        actualDataLoading = false;

        loadActualData(0);
    }

    public void firePostBodyClick(Post post) {
        if (Utils.intValueIn(post.getPostType(), VKApiPost.Type.SUGGEST, VKApiPost.Type.POSTPONE)) {
            callView(v -> v.openPostEditor(getAccountId(), post));
            return;
        }

        firePostClick(post);
    }

    public void firePostRestoreClick(Post post) {
        appendDisposable(fInteractor.restore(getAccountId(), post.getOwnerId(), post.getVkid())
                .compose(RxUtils.applyCompletableIOToMainSchedulers())
                .subscribe(dummy(), t -> callView(v -> showError(v, t))));
    }

    public void fireLikeLongClick(Post post) {
        callView(v -> v.goToLikes(getAccountId(), "post", post.getOwnerId(), post.getVkid()));
    }

    public void fireShareLongClick(Post post) {
        callView(v -> v.goToReposts(getAccountId(), "post", post.getOwnerId(), post.getVkid()));
    }

    public void fireLikeClick(Post post) {
        int accountId = getAccountId();

        appendDisposable(fInteractor.like(accountId, post.getOwnerId(), post.getVkid(), !post.isUserLikes())
                .compose(RxUtils.applySingleIOToMainSchedulers())
                .subscribe(ignore(), t -> callView(v -> showError(v, t))));
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/wallattachments/wallpostqueryattachments/WallPostQueryAttachmentsPresenter.kt
package dev.ragnarok.fenrir.fragment.wallattachments.wallpostqueryattachments

import android.os.Bundle
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.api.model.VKApiPost
import dev.ragnarok.fenrir.domain.IWallsRepository
import dev.ragnarok.fenrir.domain.Repository.walls
import dev.ragnarok.fenrir.fragment.base.PlaceSupportPresenter
import dev.ragnarok.fenrir.fromIOToMain
import dev.ragnarok.fenrir.model.*
import dev.ragnarok.fenrir.model.criteria.WallCriteria
import dev.ragnarok.fenrir.util.Utils.getCauseIfRuntime
import dev.ragnarok.fenrir.util.Utils.intValueIn
import dev.ragnarok.fenrir.util.Utils.safeCountOf
import dev.ragnarok.fenrir.util.rxutils.RxUtils.dummy
import dev.ragnarok.fenrir.util.rxutils.RxUtils.ignore
import io.reactivex.rxjava3.disposables.CompositeDisposable
import java.util.*

class WallPostQueryAttachmentsPresenter(
    accountId: Int,
    private val owner_id: Int,
    savedInstanceState: Bundle?
) : PlaceSupportPresenter<IWallPostQueryAttachmentsView>(accountId, savedInstanceState) {
    private val mPost: ArrayList<Post> = ArrayList()
    private val fInteractor: IWallsRepository = walls
    private val actualDataDisposable = CompositeDisposable()
    private var loaded = 0
    private var actualDataReceived = false
    private var endOfContent = false
    private var actualDataLoading = false
    private var Query: String? = null
    override fun onGuiCreated(viewHost: IWallPostQueryAttachmentsView) {
        super.onGuiCreated(viewHost)
        viewHost.displayData(mPost)
        resolveToolbar()
    }

    private fun loadActualData(offset: Int) {
        actualDataLoading = true
        resolveRefreshingView()
        val accountId = accountId
        actualDataDisposable.add(fInteractor.getWallNoCache(
            accountId,
            owner_id,
            offset,
            100,
            WallCriteria.MODE_ALL
        )
            .fromIOToMain()
            .subscribe({ data ->
                onActualDataReceived(
                    offset,
                    data
                )
            }) { t -> onActualDataGetError(t) })
    }

    fun fireSearchRequestChanged(q: String?, only_insert: Boolean) {
        Query = q?.trim { it <= ' ' }
        if (only_insert) {
            return
        }
        actualDataDisposable.clear()
        actualDataLoading = false
        resolveRefreshingView()
        resumedView?.onSetLoadingStatus(
            0
        )
        fireRefresh()
    }

    private fun onActualDataGetError(t: Throwable) {
        actualDataLoading = false
        showError(getCauseIfRuntime(t))
        resolveRefreshingView()
    }

    private fun doCompare(data: String?, str: List<String>): Boolean {
        data ?: return false
        for (i in str) {
            if (data.lowercase(Locale.getDefault()).contains(i)) {
                return true
            }
        }
        return false
    }

    private fun checkDocs(docs: ArrayList<Document>?, str: List<String>, ids: List<Int>): Boolean {
        if (docs.isNullOrEmpty()) {
            return false
        }
        for (i in docs) {
            if (doCompare(i.title, str) || doCompare(i.ext, str) || ids.contains(i.ownerId)) {
                return true
            }
        }
        return false
    }

    private fun checkPhotos(docs: ArrayList<Photo>?, str: List<String>, ids: List<Int>): Boolean {
        if (docs.isNullOrEmpty()) {
            return false
        }
        for (i in docs) {
            if (doCompare(i.text, str) || ids.contains(i.ownerId)) {
                return true
            }
        }
        return false
    }

    private fun checkVideos(docs: ArrayList<Video>?, str: List<String>, ids: List<Int>): Boolean {
        if (docs.isNullOrEmpty()) {
            return false
        }
        for (i in docs) {
            if (doCompare(i.title, str) || doCompare(i.description, str) || ids.contains(
                    i.ownerId
                )
            ) {
                return true
            }
        }
        return false
    }

    private fun checkAlbums(
        docs: ArrayList<PhotoAlbum>?,
        str: List<String>,
        ids: List<Int>
    ): Boolean {
        if (docs.isNullOrEmpty()) {
            return false
        }
        for (i in docs) {
            if (doCompare(i.getTitle(), str) || doCompare(i.getDescription(), str) || ids.contains(
                    i.ownerId
                )
            ) {
                return true
            }
        }
        return false
    }

    private fun checkLinks(docs: ArrayList<Link>?, str: List<String>): Boolean {
        if (docs.isNullOrEmpty()) {
            return false
        }
        for (i in docs) {
            if (doCompare(i.title, str) || doCompare(i.description, str) || doCompare(
                    i.caption, str
                )
            ) {
                return true
            }
        }
        return false
    }

    private fun checkArticles(
        docs: ArrayList<Article>?,
        str: List<String>,
        ids: List<Int>
    ): Boolean {
        if (docs.isNullOrEmpty()) {
            return false
        }
        for (i in docs) {
            if (doCompare(i.title, str) || doCompare(i.subTitle, str) || ids.contains(
                    i.ownerId
                )
            ) {
                return true
            }
        }
        return false
    }

    private fun checkPoll(docs: ArrayList<Poll>?, str: List<String>, ids: List<Int>): Boolean {
        if (docs.isNullOrEmpty()) {
            return false
        }
        for (i in docs) {
            if (doCompare(i.question ?: "", str) || ids.contains(i.ownerId)) {
                return true
            }
        }
        return false
    }

    private fun update(data: List<Post>, str: List<String>, ids: List<Int>) {
        for (i in data) {
            if (i.hasText() && doCompare(
                    i.text,
                    str
                ) || ids.contains(i.ownerId) || ids.contains(i.signerId) || ids.contains(i.authorId)
            ) {
                mPost.add(i)
            } else if (i.author != null && doCompare(i.author?.fullName, str)) {
                mPost.add(i)
            } else if (i.hasAttachments() && (checkDocs(i.attachments?.docs, str, ids)
                        || checkAlbums(
                    i.attachments?.photoAlbums,
                    str,
                    ids
                ) || checkArticles(i.attachments?.articles, str, ids)
                        || checkLinks(i.attachments?.links, str) || checkPhotos(
                    i.attachments?.photos,
                    str,
                    ids
                )
                        || checkVideos(
                    i.attachments?.videos,
                    str,
                    ids
                ) || checkPoll(i.attachments?.polls, str, ids))
            ) {
                mPost.add(i)
            }
            if (i.hasCopyHierarchy()) i.getCopyHierarchy()?.let { update(it, str, ids) }
        }
    }

    private fun onActualDataReceived(offset: Int, data: List<Post>) {
        actualDataLoading = false
        endOfContent = data.isEmpty()
        actualDataReceived = true
        if (endOfContent) resumedView?.onSetLoadingStatus(
            2
        )
        val str = (Query ?: return).split(Regex("\\|")).toTypedArray()
        for (i in str.indices) {
            str[i] = str[i].trim { it <= ' ' }.lowercase(Locale.getDefault())
        }
        val ids: MutableList<Int> = ArrayList()
        for (cc in str) {
            if (cc.contains("*id")) {
                try {
                    ids.add(cc.replace("*id", "").toInt())
                } catch (ignored: NumberFormatException) {
                }
            }
        }
        if (offset == 0) {
            loaded = data.size
            mPost.clear()
            update(data, listOf(*str), ids)
            resolveToolbar()
            view?.notifyDataSetChanged()
        } else {
            val startSize = mPost.size
            loaded += data.size
            update(data, listOf(*str), ids)
            resolveToolbar()
            view?.notifyDataAdded(
                startSize,
                mPost.size - startSize
            )
        }
        resolveRefreshingView()
    }

    public override fun onGuiResumed() {
        super.onGuiResumed()
        resolveRefreshingView()
    }

    private fun resolveRefreshingView() {
        resumedView?.showRefreshing(
            actualDataLoading
        )
        if (!endOfContent) resumedView?.onSetLoadingStatus(
            if (actualDataLoading) 1 else 0
        )
    }

    private fun resolveToolbar() {
        view?.let {
            it.toolbarTitle(getString(R.string.attachments_in_wall))
            it.toolbarSubtitle(
                getString(
                    R.string.query,
                    safeCountOf(mPost)
                ) + " " + getString(R.string.posts_analized, loaded)
            )
        }
    }

    override fun onDestroyed() {
        actualDataDisposable.dispose()
        super.onDestroyed()
    }

    fun fireScrollToEnd(): Boolean {
        if (Query.isNullOrEmpty()) {
            return true
        }
        if (!endOfContent && actualDataReceived && !actualDataLoading) {
            loadActualData(loaded)
            return false
        }
        return true
    }

    fun fireRefresh() {
        if (Query.isNullOrEmpty()) {
            return
        }
        actualDataDisposable.clear()
        actualDataLoading = false
        loadActualData(0)
    }

    fun firePostBodyClick(post: Post) {
        if (intValueIn(post.postType, VKApiPost.Type.SUGGEST, VKApiPost.Type.POSTPONE)) {
            view?.openPostEditor(
                accountId,
                post
            )
            return
        }
        firePostClick(post)
    }

    fun firePostRestoreClick(post: Post) {
        appendDisposable(fInteractor.restore(accountId, post.ownerId, post.vkid)
            .fromIOToMain()
            .subscribe(dummy()) { t ->
                showError(t)
            })
    }

    fun fireLikeLongClick(post: Post) {
        view?.goToLikes(
            accountId,
            "post",
            post.ownerId,
            post.vkid
        )
    }

    fun fireShareLongClick(post: Post) {
        view?.goToReposts(
            accountId,
            "post",
            post.ownerId,
            post.vkid
        )
    }

    fun fireLikeClick(post: Post) {
        val accountId = accountId
        appendDisposable(fInteractor.like(accountId, post.ownerId, post.vkid, !post.isUserLikes)
            .fromIOToMain()
            .subscribe(ignore()) { t ->
                showError(t)
            })
    }

}